File: benchmarks/reality.js
const fs = require('fs');

const generator = require(`${__dirname}/generator`);

const implementation = process.argv[2];
const size = process.argv[3];
const n = 1000;

const js = require('@googlemaps/polyline-codec');
const wat = require(`${__dirname}/../wat/wrapper`);
const rust = require(`${__dirname}/../rust/wrapper`);

const data = fs.readFileSync(`${__dirname}/geopuzzle.json`);
const countries = JSON.parse(data).questions;

function runBenchmark(decode, countries) {
    const start = performance.now();
    const result = countries.map((country) => {
        return {
            id: country.id,
            name: country.name,
            paths: country.polygon.map((s) => decode(s)),
        };
    });
    const end = performance.now();
    return [result, end - start];
}

(async () => {
    await wat.init();
    await rust.init();

    const decode = {
        "wat": wat.decode, "rust": rust.decode, "js": js.decode,
    }[implementation];

    let s = 0;
    for (let i = 0; i < size; i++) {
        const str = generator.generateRandomWalkPath(n);
        const result = decode(str);
        s += Math.abs(result[0][0]);
    }
    if (s >= 0) {
        const [result, time] = runBenchmark(decode, countries);
        console.log(time);
    }
})();

File: benchmarks/reality.py
import json
import random
import subprocess
import sys

from collections import defaultdict
from typing import Literal, TypeAlias

Implementation: TypeAlias = Literal["js", "rust", "wat"]


def shuffle_data_file(filename: str) -> None:
    with open(filename, "r+") as file:
        data = json.load(file)
        random.shuffle(data["questions"])

        file.seek(0)
        json.dump(data, file)


def get_algorithm_running_time(
    benchmark: str, implementation: Implementation, *args,
) -> float:
    result = subprocess.run(
        ["node", benchmark, implementation, *args], capture_output=True,
    )
    return float(result.stdout)


def collect_statistics(
    n: int, benchmark: str, filename: str,
    implementations: list[Implementation], size: int,
) -> dict[Implementation, list[float]]:
    statistics = defaultdict(list)
    for _ in range(n):
        shuffle_data_file(filename)
        for implementation in implementations:
            time = get_algorithm_running_time(
                benchmark, implementation, str(size),
            )
            statistics[implementation].append(time)
    return statistics


def main() -> None:
    sample_size = int(sys.argv[1])
    run_size = int(sys.argv[2])
    statistics = collect_statistics(
        sample_size, "reality.js", "geopuzzle.json",
        ["js", "wat", "rust"], run_size,
    )
    with open("stats/reality.json", "w") as file:
        json.dump(statistics, file)


if __name__ == "__main__":
    main()
